core/core.go

```diff
@@ -382,33 +382,76 @@ func makeAddrsFactory(cfg config.Addresses) (p2pbhost.AddrsFactory, error) {
 	}, nil
 }
 
-func makeSmuxTransport(mplexExp bool) smux.Transport {
+func makeSmuxTransport(multiplexers []string, mplexExp bool) smux.Transport {
 	mstpt := mssmux.NewBlankTransport()
 
-	ymxtpt := &yamux.Transport{
-		AcceptBacklog:          512,
-		ConnectionWriteTimeout: time.Second * 10,
-		KeepAliveInterval:      time.Second * 30,
-		EnableKeepAlive:        true,
-		MaxStreamWindowSize:    uint32(1024 * 512),
-		LogOutput:              ioutil.Discard,
+	multiplexerProtocolNames := map[string]string{
+		"yamex": "/yamux/1.0.0",
+		"mplex": "/mplex/6.7.0",
 	}
 
-	if os.Getenv("YAMUX_DEBUG") != "" {
-		ymxtpt.LogOutput = os.Stderr
+	var isYamex, isMplex bool
+
+	for _, multiplexer := range multiplexers {
+		if multiplexer == "yamex" {
+			isYamex = true
+		} else if multiplexer == "mplex" {
+			isMplex = true
+			if mplexExp {
+				log.Warning("--enable-mplex-experiment is set to false, " +
+					"but mplex is given as a multiplexer in the config.")
+			}
+		} else {
+			log.Warningf("Unknown multiplexer specified in config: %s", multiplexer)
+		}
 	}
 
-	mstpt.AddTransport("/yamux/1.0.0", ymxtpt)
+	if !isMplex && !isYamex {
+		isYamex = true
+	}
+
+	// If there are any multiplexers defined in the config, ignore mplexExp.
+	if len(multiplexers) > 0 {
+		mplexExp = false
+
+		// set order preference to match the order specified in the config
+		multiplexerProtocols := make([]string, len(multiplexers))
+		for i, multiplexer := range multiplexers {
+			if multiplexer == "yamex" || multiplexer == "mplex" {
+				multiplexerProtocols[i] = multiplexerProtocolNames[multiplexer]
+			}
+		}
 
-	if mplexExp {
-		mstpt.AddTransport("/mplex/6.7.0", mplex.DefaultTransport)
+		mstpt.OrderPreference = multiplexerProtocols
 	}
 
-	// Allow muxer preference order overriding
+	// Allow muxer preference order overriding.  If set, LIBP2P_MUX_PREFS will
+	// overwrite the order specified in the config.
 	if prefs := os.Getenv("LIBP2P_MUX_PREFS"); prefs != "" {
 		mstpt.OrderPreference = strings.Fields(prefs)
```

Comment:
Nit: can we add a `log.Error(...)` here warning the user that `LIBP2P_MUX_PREFS` is deprecated (telling them to use `Swarm.Multiplexers`).